Webauthn - Register, View, Delete and Auth

huwd opened the pull request:
[Trello](https://trello.com/c/924M5Jnv/585-webauthn-prototyping)

## What is this

This is the summation of the GOV.UK Accounts team work to add webauthn features to the acount.

Webauthn is a method of allowing a browser / device to manage a public/private key exchage in a way easy for the user to manage. In pratice it allows a pre-registered device to act as a known factor to strengthen identity assurances during login, or when we might wish to re-authorise for a high risk change (e.g. user is changing their password / email).

It is a flexible API, here we tested and produced a working demo with Yubikeys, but it could easily be extended to cover other webauthn credentails, for instance Apple's biometric IDs, Windows Hello or Android fingerprint IDs.

## What's in this

Features to:
- Register a new Webauthn Key (tested with Yubikey)
- View regsitered Webauthn Keys
- Remove registered Webauthn Keys
- Login with Webatuhn Keys

## What's missing

Development has been paused on this feature in preparation of us handing over to another team. This PR and commit messages are left to reference in the handover document to hopefully save future devs time if they are looking to impliment this in future.

Things we didn't get around to:

- Service design for fallback - Given not all browers support webauthn, and devices vary a more robust set of thinking about fallbacks when a device cannot use Webauthn is needed. The user flow currently may be unexpected. Also it is unclear if we would be satisfied with locking out a user who is using an incompatible browser, or if we always want to fall back to SMS. This impacts if Webauthn is best considered a replacement for SMS 2fa, allowing us to make mobile phone numbers optional on accounts, or if its an optional extension. We also need to think through what the preferred options are if multiple factors are registered.
- Content design - All content is placeholder
- Error messages at Registration / Authentication - Much is handled by client side javascript, meaning we don't get the benefit of our usual Rails error messaging and validation setup. There is space in the `.catch()` section of the promise chains to pick up on errors. These should be extended from a single `console.log()` to something that shows the user an error. Currently if a user attempts to authenticate with a crednetial they have not registered, things just silently fail with no warning to the user.
- A better testing strategy - we tried to impliment Feature tests (documented here) our current belief is that this isn't possible with our current stack. Instead some robust request specs and unit tests will need to fill that gap (sorry didn't get a chance to do those here!)
- JS needs to be wrapped in code that first checks for browser compatibility and sends the user down the correct path based on availble browser APIs.

## What did we learn

This is an interesting API, and becuase of its extensions it has a real opportunity to meet the user at their device and allow a broad range of 2nd factors that come "built in" (fingerprint readers, apple face ID, others).

It may not be the mostly widely used form of 2fa, but it offered excellent security and should have a place in a menu of second factor options letting the user choose what is easiest for them.

**GOV.UK**

- Bringing in client side JS node_modules is difficult at the moment. We are not using webpacker and sprokets means we would need to create our own compiled js packages to make use of high quality typescript modules currently available.
- govuk-docker may want to consider making a secure context the default dev environment if we think we will end up requiring more client side code that makes use of modern browser APIs. It will also help our dev environments look more like our prod ones.

**Anyone implimenting this**

- Base64 <-> ArrayBuffer encoding is a pain. If possible check out the [github package](https://github.com/github/webauthn-json) it would hvae made things a lot easier. Specifically ensure you use Ruby's `Base64.urlsafe_decode` for data passed in the registration callback.
- High level feature tests here are hard. We mayneed to rely on lower level tests to prove everything looks good here.

huwd committed:
Add Webauthn gem + base config initalizer

Makes your Ruby/Rails web server become a functional WebAuthn Relying
Party.

Takes care of the server-side operations needed to register or
authenticate a user credential, including the necessary cryptographic
checks.

Co-authored-by: Alan Gabbianelli <Alan.Gabbianelli@digital.cabinet-office.gov.uk>
Co-authored-by: Huw Diprose <Huw.Diprose@digital.cabinet-office.gov.uk>

huwd committed:
Add webauthn_id to user table

This will be a foreign key on each webauthn credential

Co-authored-by: Alan Gabbianelli <Alan.Gabbianelli@digital.cabinet-office.gov.uk>
Co-authored-by: Huw Diprose <Huw.Diprose@digital.cabinet-office.gov.uk>

Commit:
Add Gem Shoulda and configure configure tests

Shoulda has proved helpful for writing some quick and easy tests for
database relations using a readable syntax.

In this case it allowed us to confirm ActiveRecord associations were
correctly configured for our new user <-> credentials tables

Co-authored-by: Alan Gabbianelli <Alan.Gabbianelli@digital.cabinet-office.gov.uk>
Co-authored-by: Huw Diprose <Huw.Diprose@digital.cabinet-office.gov.uk>

Commit:
Create webauthn credentials table

The app will store a multiple webauthn credentails, identified by an
external or rawId (the same ID a browser will use to look up a
credentail in it's own store).

We will also store a Public Key, which we can use to confirm the device
has cryptographically proven they can sign data with their private key.

We also store some other niceties, like a nickname, so the user knows
which credentail is which, and a sign count which can be used to track
how many times they have sucessfully signed data with this key

Co-authored-by: Alan Gabbianelli <Alan.Gabbianelli@digital.cabinet-office.gov.uk>
Co-authored-by: Huw Diprose <Huw.Diprose@digital.cabinet-office.gov.uk>

huwd committed:
Model users and credentials

Follow on from last commit, wire the models into the database changes

Co-authored-by: Alan Gabbianelli <Alan.Gabbianelli@digital.cabinet-office.gov.uk>
Co-authored-by: Huw Diprose <Huw.Diprose@digital.cabinet-office.gov.uk>

huwd committed:
Add Webauthn Registration Feature

Adds the full feature allowing the user to register a new key.

Specific parts worth noting:

**Client side JS**

A lot of this needs to be handled client side using javascript,
here we've used a promise chain triggering closures that call functions
for each step.

There is some complexity here, mostly becuase the server-side wants
everything to be in base64Urls, and the browser API wants to recieve
everything as an ArrayBuffer. So we need to encode and decode our JSON.

There were oddities we encountered, for instance trying to pass an
ArrayBuffer straight back into JSON fails silently and leads to an empty
payload.

**Difficulty on classic GOV.UK stack bringing in NPM modules**

The complexity here could be outsourced to a NPM module. Github maintain
a well written clean module in typescript called
[Webauthn-json][webauthn-json].

Currently our GOV.UK Rails apps are still using sprokets for
compatibility with the GOV.UK publishing components. This prevented us
easily bringing these dependencies in.

Future attempts might do one of two things:

1. Keep using sprokets, but create a separate repository that takes the
github module, compiles it to plain old JS, then publishes it on NPM as
something like `gds-webauthn-json`. We could then bring that in with
node modules and sprokets.

2. Moderise our web apps to use webpacker. This is a bigger job that
would require modernising lots of JS across the platform. But this would
allow us to bring in dependencies such as bable, and add a transpile
step to our builds.

**Secure development environments**

We also encountered issues with getting access to the browser APIs in a
local dev setup. It turned out this because the `navigator.credentials`
API (core to this process) is a [secure-by-default
browser API][navigator-credentials].

The solution was to move our dev setup to a secure environment. We
explored:

- Using `mkcert` to generate local HTTPS certificates and apply them
using our docker setup or Puma. This could still be an option worth
investigating as it requires no change in URLs (leaving Plek
expectations unaffected), and brings our dev environment closer to a
12-factor app ideal of parity with production. In the end we decided to
go for a simplier route to unblock the issue more quickly.
- It was mentioned using a service that would route local dev though an
external web service that would expose it on an acutal web URL with a
real service SSL cert. However this seemed higher risk during development
and would pin us to using a new external service.
- Finally you can change URL to a reserved domain name that is pinned to
the loopback address, that the browser agrees to treat as secure. That's
what we went for here using `.localhost` in
[govuk-docker][docker-secure-env-pr]. It was quick, and worked, but
likely cannot be merged into GOV.UK docker as it breaks the expectations
of our Plek code that substitutes #{environment}.gov.uk for dev,
integration and staging.

**Difficulty feature testing**

We've left some notes in other PRs about the difficulties feature
testing this. Ideally we hoped to have one test that would cover both
our JS and ruby code from the perspective of a user (I register, now I
can see my key in security, and I can authenticate with it). We couldn't
find capybara methods that would hook into the browser to enable this.

Tests are left as notes to save some time for the next folks who try to
impliment this.

[docker-secure-env-pr]: https://github.com/alphagov/govuk-docker/pull/457
[webauthn-json]: https://github.com/github/webauthn-json
[navigator-credentials]: https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts
[secure-contexts]: https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts

Co-authored-by: Alan Gabbianelli <Alan.Gabbianelli@digital.cabinet-office.gov.uk>
Co-authored-by: Huw Diprose <Huw.Diprose@digital.cabinet-office.gov.uk>

huwd committed:
Add feature tests for view Webauthn Keys feature

These tests we could get to work, they show the null state content
renders when there are no keys.

They also show that you can view keys by nickname when you have some
registered

Co-authored-by: Alan Gabbianelli <Alan.Gabbianelli@digital.cabinet-office.gov.uk>
Co-authored-by: Huw Diprose <Huw.Diprose@digital.cabinet-office.gov.uk>

huwd committed:
Add feature tests for register Webauthn Key feature

Note these tests do not actually work, we concluded that we may not be
able to feature test this path, and that testing for this may have to
rely on lower level request specs in ruby and perhaps unit tests for the
JS. The main problem is that capybara cannot see the browser modal
popups to interact during a webauthn interaction

Left here as notes on a dead end, so that anyone trying this in future
can more quickly reach a better testing outcome.

Co-authored-by: Alan Gabbianelli <Alan.Gabbianelli@digital.cabinet-office.gov.uk>
Co-authored-by: Huw Diprose <Huw.Diprose@digital.cabinet-office.gov.uk>

huwd committed:
Add Webatuhn View keys Feature

Lets the user see which keys have been registered

Co-authored-by: Alan Gabbianelli <Alan.Gabbianelli@digital.cabinet-office.gov.uk>
Co-authored-by: Huw Diprose <Huw.Diprose@digital.cabinet-office.gov.uk>

huwd committed:
Add Webauthn Delete keys Feature

Allows the user to remove keys registered with their account

Co-authored-by: Alan Gabbianelli <Alan.Gabbianelli@digital.cabinet-office.gov.uk>
Co-authored-by: Huw Diprose <Huw.Diprose@digital.cabinet-office.gov.uk>

huwd committed:
Add Webauthn Authenticate at Login feature

Very similar challenges / code to the registration feature but in this
case we are using the navigator.credentails.get featre and expecting
slightly different data to be passed back and forth.

Co-authored-by: Alan Gabbianelli <Alan.Gabbianelli@digital.cabinet-office.gov.uk>
Co-authored-by: Huw Diprose <Huw.Diprose@digital.cabinet-office.gov.uk>

huwd committed:
Configure Webauthn to expect base64url encoding

We ran into some encoding expectation issues when trying to parse
external IDs during registration and authentication. The solution was to
switch from using `Base64.strict_encode` to `Base64.urlsafe_encode` with
the `padding: true` option.

Here we can additionally put this expectation explicitly into the
configuration of the app.

Co-authored-by: Alan Gabbianelli <Alan.Gabbianelli@digital.cabinet-office.gov.uk>
Co-authored-by: Huw Diprose <Huw.Diprose@digital.cabinet-office.gov.uk>

huwd commented:
@jonathanglassman would you be up for reviewing the PR text and perhaps even the commit messages?
I'd like these to end up as almost pseudo-docs we could leave to our digi-ID friends, hopefully this spike and what we learned could be useful for them in the future?

@AlanGabbianelli shall we both do a quick self-review make sure we've not left anything silly in there, or I mashed something during our rebase?

then @barrucadu would be interested in thoughts you have about any of the approach here that anyone else might want to consider in future?

huwd committed:
TOfixup: auth tests

barrucadu commented on config/initializers/webauthn.rb:
```diff
@@ -0,0 +1,35 @@
+WebAuthn.configure do |config|
+  # This value needs to match `window.location.origin` evaluated by
+  # the User Agent during registration and authentication ceremonies.
+  config.origin = "http://www.login.service.localhost"
```
A lot of these APIs need to know the hostname they're running on, which I guess makes sense but is also a potential source of bugs - we use `ENV["OIDC_IDP_ISSUER"] || ENV["REDIRECT_BASE_URL"]` for this in the OIDC initialiser.

huwd replied:
Yeah great point, also worth nothing this is one of the config bits that means this PR would not be ready for game-time, further polish would be required.

NB there's also the option to make the relaying party differ from the default.
The gem author leaves comments [here](https://github.com/alphagov/govuk-account-manager-prototype/pull/704/files/6b1c124feed46af49bb97208cd4077f3776fb545#diff-bc7d5d6cad7fa7aef1230014f824a1f497e64bf498e00f47dde966801a682b04R16-R18) and [here](https://github.com/alphagov/govuk-account-manager-prototype/pull/704/files/6b1c124feed46af49bb97208cd4077f3776fb545#diff-bc7d5d6cad7fa7aef1230014f824a1f497e64bf498e00f47dde966801a682b04R20-R23)

barrucadu commented on db/migrate/20210203095633_add_webauthn_id_to_user.rb:
```diff
@@ -0,0 +1,5 @@
+class AddWebauthnIdToUser < ActiveRecord::Migration[6.0]
+  def change
+    add_column :users, :webauthn_id, :string
```
Is this set up so a user can only have one WebAuthn credential?

If we were doing this for real we'd want to support multiple credentials per account, as the credentials (can be) device specific.

huwd replied:
Yeah so this does support multiple credentails.
I named this `webauthn_id` because that's what the ruby library has it as. I think we could rename to clarify though.

This is more like the `user_webauthn_id`, it identfies a user not a credential.

barrucadu replied:
Ah, I see, this is a ["user handle"](https://www.w3.org/TR/webauthn-2/#user-handle) - a user ID in webauthn-speak.

huwd replied:
yeah that'd be a much better name for it if I was doing this again

barrucadu commented on spec/models/webauthn_credential_spec.rb:
```diff
@@ -0,0 +1,19 @@
+RSpec.describe WebauthnCredential do
```
General point: I don't think we should test that factories are valid - that should be covered (indirectly) by other tests which use the factories.

huwd replied:
@AlanGabbianelli and I had a chat about this one of our first days pairing.
We'd spoken about how in other repos we'd not done this within a model test, but rather had a separate test that just iterates over all factories and variations in traits and tests for validity. That should be less code and a better genrealised test than what we did here.

I do think there's value in it though, I could imagine scenarios where we've run a migration change and not updated the factories. Ideally that would get picked up in the tests like you say, but possible not?

And these are essentially low cost, quick unit tests, so I don't see much harm?

huwd committed:
TOfixup: remove unused placeholder content

barrucadu commented on app/assets/javascripts/webauthn-registration.js:
```diff
@@ -0,0 +1,106 @@
+var form = document.querySelector('.webauthn_key_registration')
```
Needing to translate the GitHub TS to JS is a pain... and also a potential copyright worry.

From the lofty position of a backend developer who wouldn't do any of the work, I think the best solution would be for the GOV.UK frontend stuff to switch to webpack.  It would probably be a lot of work, but it's the way the JS ecosystem has been moving for a while now, and TS is only getting more common.

On the other hand, that's not going to happen for a while.  So maybe we do just have to write our own WebAuthn client-side implementation.

This could even be a point against using Rails / govuk_publishing_components.  There are service teams using the Design System outside of Ruby.

huwd replied:
ah super interestng, Hadn't considered the copyright angle 😬.
Yeah i 💯 agree!

huwd committed:
tofixup: view key feature spec - shoulda been plural

barrucadu commented on app/controllers/webauthn_registration_controller.rb:
```diff
@@ -0,0 +1,63 @@
+    end
+  end
+
+  def generate_random_nickname
+      letters = [('a'..'z'), ('A'..'Z')].map(&:to_a).flatten
+      return (0...10).map { letters[rand(letters.length)] }.join
+  end
```
If we were doing this for real, I think generating the name client-side (so the user can see it) and using three or four English words (like the default GitHub repo names) would be much better.

huwd replied:
Yeah words sounds great - much more memorable.
And agreed, if it were client side we could simply this!

barrucadu commented on config/routes.rb:
```diff
@@ -26,10 +26,19 @@
+            get "/options", to: "webauthn_registration#create", as: :webauthn_register_request_options
+            post "/callback", to: "webauthn_registration#callback", as: :webauthn_register_callback
+          end
+        end
```
Nitpick: the registration endpoints should go under `/edit`, and we'd also need something under `/mfa` to let users authenticate mid-session if they'd previously skipped.

huwd committed:
TOFIXUP: normalise security content

huwd committed:
TOFIXUP: normalise webauthn content

huwd committed:
TOFIXUP: normalise MFA content

barrucadu commented on app/controllers/sessions_controller.rb:
```diff
@@ -44,8 +48,12 @@ def create
+          MultiFactorAuth.generate_and_send_code(resource)
+          redirect_to enter_phone_code_path
+        end
       else
```
I think we noticed that this doesn't work when MFA is skipped

huwd replied:
Yeah this needs more robust testing and a bit of service design I think.
Right at the end I think I managed to demo to Alan that it did skip MFA if we:
- started with no keys, logged in with SMS 2FA, checking remember me
- registered a key
- logged out
- logged back in, it went straight to account.

I think we'd want to consider if we add a similar remember me box to the webauth, or if that gets ignored for login / other auth points? And if so when  / where we should clear the cookie.

These fallbacks should probably also sit in some kind of preferred heirarchy (like if we have webauthn, do we always prefer to that SMS or OTP?) - then we would have looked to package up the Remember Me feature inside that new heirarchy? (what happens if you add a new factor... do we still remember you, or re-challenge?)

barrucadu commented:
Looks generally good to me, obviously this is a spike, and not a feature ready to ship, so I've not done a very thorough review.

I think it's taught us useful things:

1. We need to make sure that we gracefully fall back to a different factor if a user doesn't have a browser with WebAuthn support (or if they're using a different device)
2. Using pre-existing WebAuthn client-side libraries is a big pain with how javascript current works on GOV.UK
3. There's a fair bit of verbosity in adding a new MFA method to our app: there's potentially four parts to implement, with slightly different (but very similar) code in each:
    - Using MFA to log in
    - Setting up MFA at registration
    - Changing an MFA method once logged in
    - Re-authenticating after previously skipping MFA

I think (2) is maybe something we can take back to the GOV.UK developer community.  Perhaps we could talk about this spike at a Tech Fortnightly?  It's worrying if we're locking ourselves out of big chunks of the JS ecosystem by not supporting webpack.

And (3) is definitely a lesson for Digital Identity to take forward: make it easy to implement new MFA methods, try not to have different-but-very-similar code everywhere where MFA crops up.

huwd committed:
Appease the linter

Should probably rebase this back in.. some of the ruby in prior commits didn't match our styleguide

huwd committed:
Appease the linter

Should probably rebase this back in.. some of the ruby in prior commits didn't match our styleguide

huwd commented:
really nice summary, thanks!

huwd committed:
Defensively wrap addEventListener

The JS is added globally and nearly all pages will not have the forms we are targeting

We should only try to run the addEventListener if the form is not null

jonathanglassman commented:
> @jonathanglassman would you be up for reviewing the PR text and perhaps even the commit messages?
> I'd like these to end up as almost pseudo-docs we could leave to our digi-ID friends, hopefully this spike and what we learned could be useful for them in the future?
> 
> @AlanGabbianelli shall we both do a quick self-review make sure we've not left anything silly in there, or I mashed something during our rebase?
> 
> then @barrucadu would be interested in thoughts you have about any of the approach here that anyone else might want to consider in future?

@huwd I've had a look at the commit messages and the PR description. Looks ok to me in general. My biggest question is how does someone from DI who wants to implement this into their project or app actually do so? Do we need to tell them? There's a lot of information from this spike, but nothing that I could see saying "This is how you:

- register a new Webauthn Key (tested with Yubikey)
- view regsitered Webauthn Keys
- remove registered Webauthn Keys
- login with Webatuhn Keys"

If it is in the content and I haven't seen it, or it's very obvious to the intended developer audience and does not need any further content saying "this is is how you do the thing", or that wasn't the point of this spike, then fair enough.

huwd commented:
Closing this down now.
